packages/block-editor/src/components/provider/index.js

```diff
@@ -34,10 +34,21 @@ class BlockEditorProvider extends Component {
 			this.attachChangeObserver( registry );
 		}
 
-		if ( this.isSyncingOutcomingValue ) {
-			this.isSyncingOutcomingValue = false;
+		if ( this.isSyncingOutcomingValue !== null && this.isSyncingOutcomingValue === value ) {
+			// Skip block reset if the value matches expected outbound sync
+			// triggered by this component by a preceding change detection.
+			// Only skip if the value matches expectation, since a reset should
+			// still occur if the value is modified (not equal by reference),
+			// to allow that the consumer may apply modifications to reflect
+			// back on the editor.
+			this.isSyncingOutcomingValue = null;
 		} else if ( value !== prevProps.value ) {
-			this.isSyncingIncomingValue = true;
+			// Reset changing value in all other cases than the sync described
+			// above. Since this can be reached in an update following an out-
+			// bound sync, unset the outbound value to avoid considering it in
+			// subsequent renders.
+			this.isSyncingOutcomingValue = null;
+			this.isSyncingIncomingValue = value;
```

Comment:
I know these changes are really tricky and fragile. The good news is that we have e2e tests for most of the use-cases and you can endup fixing one but breaking others :P